server/db/db.js

```diff
@@ -325,6 +330,64 @@ module.exports.getInstallationWithInstallationId = async function(installationId
     return null
 }
 
+module.exports.getHeartbeats = async function(client) {
+    let transactionMode = (typeof client !== 'undefined')
+    if(!transactionMode) {
+        client = await pool.connect()
+    }
+    
+    const { rows } = await client.query("SELECT * FROM heartbeat")
+    
+    if(!transactionMode) {
+        client.release()
+    }
+    
+    if(rows.length > 0) {        
+        return rows.map(createHeartbeatFromRow)
+    }
+    return []
+}
+
+module.exports.getHeartbeatWithSystemId = async function(systemId, client) {
+    let transactionMode = (typeof client !== 'undefined')
+    if(!transactionMode) {
+        client = await pool.connect()
+    }
+    
+    let { rows } = await client.query("SELECT * FROM heartbeat WHERE system_id = $1", [systemId])
+    
+    if(!transactionMode) {
+        client.release()
+    }
+    
+    if(rows.length > 0) {
+        return createHeartbeatFromRow(rows[0])
+    }
+    return null
+}
+
+module.exports.saveHeartbeat = async function(heartbeat, client) {
+    let transactionMode = (typeof client !== 'undefined')
+    if(!transactionMode) {
+        client = await pool.connect()
+    }
+    
+    let { rows } = await client.query("SELECT * FROM heartbeat WHERE system_id = $1 LIMIT 1", [heartbeat.systemId])
+    if(rows.length === 0) {
+        if(!transactionMode) {
+            client.release()
+        }
+        throw new Error("Tried to save a heartbeat that doesn't exist yet.")
+    }
+    const query = "UPDATE heartbeat SET system_id = $1, flic_last_seen_time = $2, flic_last_ping_time = $3, heartbeat_last_seen_time = $4, system_name = $5, hidden = $6, sent_alerts = $7, muted = $8, twilio_alert_number = $9, heartbeat_alert_recipients = $10 WHERE system_id = $11"
```

Comment:
I'm concerned that because we're updating everything we might end up with a problem if more than one change is going on at a time. For example, if the RPi calls `/heartbeat` at the same time that I try to `mute` an installation, it's possible that they both retrieve from the database, getting the same values, and then when they save, only the second one that saves will have its data written.

Less of a big deal if the `/heartbeat` data is overwritten since it's being called every second. But we could get unlucky with the muted one and have to call it more than once before it is effective.

I think we should only be updating the value that changed, not all of them each time. Or putting the read and write into the same transaction to keep it sane.